Add client user connection limit

AndrewJackson2020 opened the pull request:
This PR adds a user level client connection limit. I believe that currently user connections can only be limited by how many people are given connections  to the database backends but there is no way to limit the number of connections to pgbouncer itself. Combined with the global connection limit this means that one user could use up all connections on the pgbouncer instance. This PR proposes a fix to this problem.

Related to #1136 (because it partially fixes it)

JelteF commented on include/bouncer.h:
```diff
@@ -540,7 +540,9 @@ struct PgGlobalUser {
 	int pool_mode;
 	int pool_size;				/* max server connections in one pool */
 	int max_user_connections;	/* how much server connections are allowed */
+	int max_user_client_connections;	/* how much server connections are allowed */
```
```suggestion
	int max_user_client_connections;	/* how many server connections are allowed */
```

AndrewJackson2020 replied:
Done

JelteF commented on include/bouncer.h:
```diff
@@ -540,7 +540,9 @@ struct PgGlobalUser {
 	int max_user_connections;	/* how much server connections are allowed */
+	int max_user_client_connections;	/* how much server connections are allowed */
 	int connection_count;	/* how much connections are used by user now */
+	int client_connection_count;	/* how much connections are used by user now */
```
```suggestion
	int client_connection_count;	/* how many client connections are used by user now */
```

AndrewJackson2020 replied:
Done

JelteF commented on include/bouncer.h:
```diff
@@ -540,7 +540,9 @@ struct PgGlobalUser {
 	int pool_size;				/* max server connections in one pool */
 	int max_user_connections;	/* how much server connections are allowed */
+	int max_user_client_connections;	/* how much server connections are allowed */
 	int connection_count;	/* how much connections are used by user now */
```
```suggestion
	int connection_count;	/* how many server connections are used by user now */
```

AndrewJackson2020 replied:
Done

JelteF commented on src/client.c:
```diff
@@ -489,6 +504,8 @@ bool set_pool(PgSocket *client, const char *dbname, const char *username, const
 				if (!client->db->auth_user_credentials)
 					client->db->auth_user_credentials = add_global_credentials(cf_auth_user, "");
 			}
+
+
```
Can you undo this?

AndrewJackson2020 replied:
Done

JelteF reviewed (changes requested):
This seems like a good solution to the described problem. A similar setting would probably be useful per database, but that doesn't need to be in this PR. 

Some things that should be changed about the PR:
1. Github Actions should be green (fix C warning and fix formatting)
2. A test should be added (`test/test_limits.py` seems like a good place to add one)
3. The new option should be documented in `doc/config.md`

eulerto commented:
I didn't review this PR yet but I noticed that you did lots of changes in the documentation that is not related to this feature. If you want to propose a whitespace cleanup, create a separate PR. We generally use `/* comment */` instead of `// comment`.

BTW, I'm fine with this feature. It is equivalent to `CREATE ROLE ... CONNECTION LIMIT`.

AndrewJackson2020 commented:
> I didn't review this PR yet but I noticed that you did lots of changes in the documentation that is not related to this feature. If you want to propose a whitespace cleanup, create a separate PR. We generally use `/* comment */` instead of `// comment`.
> 
> BTW, I'm fine with this feature. It is equivalent to `CREATE ROLE ... CONNECTION LIMIT`.

Apologies about that, my editor eliminates trailing whitespace default. Will revert momentarily.

AndrewJackson2020 commented:
> I didn't review this PR yet but I noticed that you did lots of changes in the documentation that is not related to this feature. If you want to propose a whitespace cleanup, create a separate PR. We generally use `/* comment */` instead of `// comment`.
> 
> BTW, I'm fine with this feature. It is equivalent to `CREATE ROLE ... CONNECTION LIMIT`.

I reverted the whitespace removal and changed the comment style to conform with the rest of the codebase.

JelteF reviewed (changes requested):
Thanks! That looks a lot better already. I left a review just now on your other #1138 PR.  Most of the feedback there should also be applied here. Other than that it looks pretty good.

I still have want to do some manual tests of both PRs though

AndrewJackson2020 commented:
@JelteF Made all of the documentation changes and added a test of the stats command in the unit tests. Also split the unit test out into 2 separate ones: one that is meant to deny a connection and one that is meant to accept a connection, I feel this is a bit cleaner.

A few additional thoughts:
1. Should this PR and #1138 be merged because they are so similar and have similar feedback?
2. Should the name of the variables be changed from max_user_client_connections/max_user_connections to max_user_client_connections/max_user_server_connections?
3. Should the tests of the stats command  be moved to test/test_admin.py or is it fine as is?

JelteF commented on src/objects.c:
```diff
@@ -1393,6 +1393,11 @@ void disconnect_server(PgSocket *server, bool send_term, const char *reason, ...
 {
+	if (client->login_user_credentials) {
+		if (client->login_user_credentials->global_user) {
+			client->login_user_credentials->global_user->client_connection_count--;
```
This substraction should only be done when this specific clien actually increased the connection count before. 

Right now if a user is at the limit. The next failing connection attempt (due to hitting the limit) will then also decrease the connection count. So the connection attempt after that will succeed again, eventhough we're still at the client connection limit.

AndrewJackson2020 replied:
Okay so I added a unit test assert involving the situation you described above. Moved the increment to right before the initial check and changed the inequality accordingly. Seems to satisfy all test cases though there might be another corner case I'm not seeing.

An alternate way of going about this is adding a boolean to each client that specifies if they contribute to the user level count and check that before decrementing the count. I feel like the logic in this commit is better since it adds less variables, as long as the logic is correct.

Another thing that I notice is that this new logic does not apply to PAM authed users though I feel like it should.

JelteF replied:
> An alternate way of going about this is adding a boolean to each client that specifies if they contribute to the user level count and check that before decrementing the count. I feel like the logic in this commit is better since it adds less variables, as long as the logic is correct.

Agreed.

> Another thing that I notice is that this new logic does not apply to PAM authed users though I feel like it should.

Agreed. Everywhere we set `client->login_user_credentials` we should do this incrementing/decrementing correctly. Also the block below where auth_user is used should do this. I think the easiest way to do this is do

```
client->login_user_credentials = /* something */
if (!check_user_connection_count())
    return;

bool check_user_connection_count() {
    /* your check */
}
```

AndrewJackson2020 replied:
Encapsulated the logic into the function and added the call to everywhere client->login_user_credentials is assigned. Tests all passed but the surface area of the change is now much bigger which may mean more testing is required (limits correctly incremented, decremented and enforced for pam users, admin users, etc). Also the check_user_connection_count function could probably be refactored a bit (lot of nesting).

AndrewJackson2020 replied:
@JelteF Expanded test coverage to include admin users. Refactored the check_user_connection_count function to prefer early returrns instead of nested if statements. Refactored the tests so they should be faster, rely less on asynchronously running pg_sleep.

Could also expand test coverage to stats users as I believe are instantiated differently than admin users. Also missing is test coverage around PAM users though I dont believe there is PAM coverage elsewhere in the test codebase either.

Let me know if anything else needs to be done with this PR in order to get it in a mergable state.

JelteF replied:
If tests pass, I think this is probably good. I'm unavailable for the next ~3 weeks though.

AndrewJackson2020 replied:
Understood, thanks for the heads up.

JelteF commented:
>     1. Should this PR and [Add db client connection limit #1138](https://github.com/pgbouncer/pgbouncer/pull/1138) be merged because they are so similar and have similar feedback?


>     2. Should the name of the variables be changed from max_user_client_connections/max_user_connections to max_user_client_connections/max_user_server_connections?

Let's keep them as is. I don't think the better naming (because I agree it is a bit confusing now) is worth breaking backwards compatibility of the config setting names.

>     3. Should the tests of the stats command  be moved to test/test_admin.py or is it fine as is?

It's fine as is imho.

JelteF commented on src/client.c:
```diff
@@ -476,6 +478,22 @@ bool set_pool(PgSocket *client, const char *dbname, const char *username, const
+
+		/* Check client_connection count limit */
+		if (client->login_user_credentials) {
+			client->login_user_credentials->global_user->client_connection_count++;
```
Let's add a comment here for clarity.
```suggestion
			/* increment count now, so that we can decrement it safely in disconnect_client if limit was reached */
			client->login_user_credentials->global_user->client_connection_count++;
```

JelteF reviewed (changes requested):
Sorry for the delay in reviewing. I think this is almost good now, some final suggested changes.

AndrewJackson2020 commented on src/admin.c:
```diff
@@ -1676,6 +1680,9 @@ bool admin_pre_login(PgSocket *client, const char *username)
 		if (res >= 0 && peer_uid == getuid()
 		    && strcmp("pgbouncer", username) == 0) {
 			client->login_user_credentials = admin_pool->db->forced_user_credentials;
+			if (!check_user_connection_count(client)) {
```
Good point. Looking at the implementation of the server wide max_client_conn setting it behaves in the way you are describing. Will update that.

AndrewJackson2020 replied:
This is fixed now. Added unit tests to verify that the connection limit applies to the stats users accessing the admin console but not admin users accessing the admin console.

AndrewJackson2020 commented on doc/config.md:
```diff
@@ -1420,7 +1432,7 @@ client connections, if a password-based authentication method is
 connections to the backend server, if the backend server requires
 password-based authentication (unless the password is specified
-directly in the database's connection string).
+directly in the database's connection string). 
```
Sorry about that, will fix.

AndrewJackson2020 commented on src/client.c:
```diff
@@ -504,6 +545,9 @@ bool set_pool(PgSocket *client, const char *dbname, const char *username, const
 
 			slog_info(client, "no such user: %s", username);
 			client->login_user_credentials = calloc(1, sizeof(*client->login_user_credentials));
+			if (!check_user_connection_count(client)) {
+				return false;
+			}
```
Will do. Will also make sure test coverage for this feature includes users from the auth DB.

AndrewJackson2020 commented:
@JelteF Pushed a new commit that fixes a bunch of the issues the issues that you mentioned. 

Found that the auth pass through issue that you mentioned in #1138 applies to this PR as well. Implemented a somewhat hacky fix to this. Would appreciate your feedback on this or if you have any alternative ideas.

Also expanded unit test coverage for stats users and auth pass through users. One area that isnt covered is PAM but looking through the unit tests I dont think there is any test coverage of PAM currently so I think adding that here might be out of scope for this PR though it would be very beneficial.

AndrewJackson2020 commented:
@JelteF Fixed the macOS tests by scaling back the run_with_config usage as described in #1138.

Let me know if this solution works for you or if it needs to be reworked more.

AndrewJackson2020 commented:
@JelteF Anything else I can do here to help this PR get merged?

JelteF commented:
No, it's blocked on me finding some time to review PgBouncer PRs again. Busy busy.

JelteF commented on src/client.c:
```diff
@@ -420,6 +420,42 @@ static bool finish_set_pool(PgSocket *client, bool takeover)
+	if (!client->login_user_credentials)
+		return true;
+
+	/* increment count now, so that we can decrement it safely in disconnect_client if limit was reached */
```
That comment seems out of date now.
```suggestion
```

JelteF commented on src/client.c:
```diff
@@ -420,6 +420,41 @@ static bool finish_set_pool(PgSocket *client, bool takeover)
+	if (!client->login_user_credentials->global_user)
+		return true;
+
+	if (!client->user_connection_counted) {
```
I agree this is not pretty. But given the amount of counting bugs found in earlier versions, I prefer this way because this one is obviously correct and will stay correct in the future. I added a check for user_connection_counted in the decrement path too.

AndrewJackson2020 commented:
@JelteF anything else you want done with this PR before merging?

JelteF commented:
I don't remember why I didn't press merge yet, maybe because CI was failing for unrelated reasons.
